woocommerce/woocommerce-ios
# [IAP] Check whether transactions in `Transaction.Updates` queue have been handled

Closes: #10187 


## Description
This PR handles In-App Purchases transactions that are currently in the `Transaction.Updates` queue. These transactions could be of several types, for example incomplete purchases, renewals, or switching between plans... and can happen both in-app, or outside the app via the customer App Store. 

When we run the app, we'll be listening to updates to these transactions and handle these accordingly.

Context: pdfdoF-3hA-p2

## Changes:
In the IAP transaction listener, we're now checking for verified transactions, and when we got one in the updates queue, we attempt to reach the `iap/transactions/` endpoint with that transaction ID.

The remote will either return a value (transaction already exists in WPCOM) or throw an unhandled error (transaction does not exist in WPCOM).

- If exists, we just mark the transaction as `.finish()`
- If does not exist, the catch block will be triggered and we'll perform the previously existing `handleCompletedTransaction(:)`

## Testing instructions
Testing this has proven to be tricky, happy to get any feedback on the testing instructions or if there's a better way to do it with Apple built-in StoreKit tools. For now, this seems to work fine:

1. Add a breakpoint in [line 297](https://github.com/woocommerce/woocommerce-ios/blob/1bd40b781bef2f84881fa5939790102488772ce6/Yosemite/Yosemite/Stores/InAppPurchaseStore.swift#L297) and [line 302](https://github.com/woocommerce/woocommerce-ios/blob/1bd40b781bef2f84881fa5939790102488772ce6/Yosemite/Yosemite/Stores/InAppPurchaseStore.swift#L302), so we can stop in the `do` or `catch` blocks depending on the case we're testing. 

2. Using the `WooCommerce IAP` scheme, on a Free Trial site eligible for IAP (a new site can be created in Menu > Switch Sites > Add New Site)
3. In Xcode, open `Debug > StoreKit > Manage Transactions`

<img width="1546" alt="Screenshot 2023-07-19 at 18 51 29" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/7857a7ef-1c53-4950-aa3c-f587d53efec7">

4.  Tap on `Upgrade Now` > select a plan > and see an error (expected since is the testing target and we're using a simulator), and the transaction appears on the Manage Transactions.

<img width="1425" alt="Screenshot 2023-07-19 at 18 52 08" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/30ea8c84-068c-4331-a89f-124e6f7e2120">

5. Re-run the app so the transaction listener is triggered.  Observe that since this is a non existing transaction in the WPCOM billing system, we'll land in the `catch` block, and `handleCompletedTransaction(result)` will be called.

6. In `Debug > StoreKit > Manage Transactions` delete any existing transaction by tapping the "trash" icon, so we can clear the state.

7. In [line 296](https://github.com/woocommerce/woocommerce-ios/blob/1bd40b781bef2f84881fa5939790102488772ce6/Yosemite/Yosemite/Stores/InAppPurchaseStore.swift#L296) change the current `transaction.id` for `UInt64(2000000370069639)`. This is a real transaction that has already been handled in WPCOM billing system:

```diff
	do {
		_ = await self.pauseTransactionListener.values.contains(false)
		// For verified transactions, check whether a transaction has been handled already,
		// and mark it as finished if that's the case, or handle it otherwise
+		let someRealTransactionID = UInt64(2000000370069639)
+		_ = try await self.remote.retrieveHandledTransactionSiteID(for: someRealTransactionID)
-		_ = try await self.remote.retrieveHandledTransactionSiteID(for: transaction.id)
		await transaction.finish()

		self.logInfo("Marking transaction \(transaction.id) as finished")
	}
```
8.  Since transactions are clear, we can upgrade again. Tap on `Upgrade Now` > select a plan > and see an error (expected since is the testing target and we're using a simulator), and a transaction appears on the Manage Transactions.

9. Re-run the app so the transaction listener is triggered.  Observe that since now we have transaction in the queue, but we switched the transaction ID, we'll land in the `do` block, we reach the `retrieveHandledTransactionSiteID` remote and the transaction will be marked as finished since already exists in our end.

10. Note that the console will display the "wrong" transaction ID, since the one printed is from the test transaction we performed through the app, but the one that actually made the flow finish was the one we've injected.
```
2023-07-19 18:52:32.562436+0200 WooCommerce[88904:9172591] [💰IAP Store] Marking transaction 0 as finished
```
